src/main/java/com/griddynamics/cloud/learning/dal/repositories/Repository.java

```diff
@@ -0,0 +1,63 @@
+package com.griddynamics.cloud.learning.dal.repositories;
+
+import com.griddynamics.cloud.learning.dal.connection.DBConnectionManager;
+import com.griddynamics.cloud.learning.dal.domain.DomainEntity;
+
+import java.sql.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+public abstract class Repository<T extends DomainEntity> {
+
+    private static final Logger logger = Logger.getLogger(Repository.class.getName());
+
+    protected abstract String getEntityTableName();
+
+    protected abstract Optional<T> convertToEntity(ResultSet resultSet);
+
+    public Optional<T> getById(long id) {
+        String query = "SELECT * FROM " + getEntityTableName() + " WHERE id = " + id + ";";
+        return getEntityByParameter(query);
+    }
+
+    public Optional<List<T>> getAll(int pageNumber, int limitPerPage) {
+        String query = "SELECT * FROM " + getEntityTableName();
+        String paginatedQuery = paginate(query, pageNumber, limitPerPage);
+        return getArticles(paginatedQuery);
+    }
+
+    protected Optional<T> getEntityByParameter(String query) {
+        try (Connection connection = DBConnectionManager.getConnection();
+             PreparedStatement statement = connection.prepareStatement(query)) {
+            ResultSet resultSet = statement.executeQuery();
+            resultSet.next();
+            return convertToEntity(resultSet);
+        } catch (SQLException e) {
+            logger.log(Level.SEVERE, e.getMessage(), e);
+        }
+        return Optional.empty();
+    }
+
+    protected Optional<List<T>> getArticles(String query) {
+        try (Connection connection = DBConnectionManager.getConnection();
+             PreparedStatement statement = connection.prepareStatement(query)) {
+            ResultSet resultSet = statement.executeQuery();
+            List<T> entities = new ArrayList<>();
+            while (resultSet.next()) {
+                Optional<T> entity = convertToEntity(resultSet);
+                entity.ifPresent(entities::add);
+            }
+            return Optional.of(entities);
+        } catch (SQLException e) {
+            logger.log(Level.SEVERE, e.getMessage(), e);
+        }
+        return Optional.empty();
+    }
+
+    protected String paginate(String query, int pageNumber, int limitPerPage) {
```

Comment:
1) The result of our pagination should also contain total rows count.
2) Why do we have order by id here? First of all it should be a part of the request, because we might have different fields fo sorting. Plus we also have to handle sorting direction.